File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/045-FLOR.Flora.swift
//
//  FLORRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class FLORRecord: Record, IHaveEDID, IHaveMODL {
    public override var description: String { return "FLOR: \(EDID)" }
    public var EDID: STRVField = STRVField_empty // Editor ID
    public var MODL: MODLGroup? = nil // Model
    public var FULL: STRVField! // Plant Name
    public var SCRI: FMIDField<SCPTRecord>? = nil // Script (optional)
    public var PFIG: FMIDField<INGRRecord>? = nil // The ingredient the plant produces (optional)
    public var PFPC: BYTVField! // Spring, Summer, Fall, Winter Ingredient Production (byte)
    
    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID": EDID = r.readSTRV(dataSize)
        case "MODL": MODL = MODLGroup(r, dataSize)
        case "MODB": MODL!.MODBField(r, dataSize)
        case "MODT": MODL!.MODTField(r, dataSize)
        case "FULL": FULL = r.readSTRV(dataSize)
        case "SCRI": SCRI = FMIDField<SCPTRecord>(r, dataSize)
        case "PFIG": PFIG = FMIDField<INGRRecord>(r, dataSize)
        case "PFPC": PFPC = r.readBYTV(dataSize)
        default: return false
        }
        return true
    }
}

File: src/ObjectManager/Object.Tes/FilePacks/Records/045-FLOR.Flora.cs
using OA.Core;

namespace OA.Tes.FilePacks.Records
{
    public class FLORRecord : Record, IHaveEDID, IHaveMODL
    {
        public override string ToString() => $"FLOR: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public MODLGroup MODL { get; set; } // Model
        public STRVField FULL; // Plant Name
        public FMIDField<SCPTRecord> SCRI; // Script (optional)
        public FMIDField<INGRRecord> PFIG; // The ingredient the plant produces (optional)
        public BYTVField PFPC; // Spring, Summer, Fall, Winter Ingredient Production (byte)

        public override bool CreateField(UnityBinaryReader r, GameFormatId format, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID": EDID = new STRVField(r, dataSize); return true;
                case "MODL": MODL = new MODLGroup(r, dataSize); return true;
                case "MODB": MODL.MODBField(r, dataSize); return true;
                case "MODT": MODL.MODTField(r, dataSize); return true;
                case "FULL": FULL = new STRVField(r, dataSize); return true;
                case "SCRI": SCRI = new FMIDField<SCPTRecord>(r, dataSize); return true;
                case "PFIG": PFIG = new FMIDField<INGRRecord>(r, dataSize); return true;
                case "PFPC": PFPC = new BYTVField(r, dataSize); return true;
                default: return false;
            }
        }
    }
}